In [1]:
import tensorflow as tf

# Helper libraries
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.3.0


In [2]:
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

In [3]:
train_images = train_images / 255.0

test_images = test_images / 255.0

## Model #1:  Original Model -  1. adam, dense128, dense10

In [4]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [5]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [6]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 603us/step - loss: 0.4932 - accuracy: 0.8274
Epoch 2/10
1875/1875 [==============================] - 1s 675us/step - loss: 0.3767 - accuracy: 0.8646
Epoch 3/10
1875/1875 [==============================] - 1s 689us/step - loss: 0.3387 - accuracy: 0.8784
Epoch 4/10
1875/1875 [==============================] - 1s 719us/step - loss: 0.3136 - accuracy: 0.8849
Epoch 5/10
1875/1875 [==============================] - 1s 652us/step - loss: 0.2968 - accuracy: 0.8911
Epoch 6/10
1875/1875 [==============================] - 1s 650us/step - loss: 0.2794 - accuracy: 0.8965
Epoch 7/10
1875/1875 [==============================] - 1s 655us/step - loss: 0.2685 - accuracy: 0.9014
Epoch 8/10
1875/1875 [==============================] - 1s 637us/step - loss: 0.2585 - accuracy: 0.9043
Epoch 9/10
1875/1875 [==============================] - 1s 648us/step - loss: 0.2479 - accuracy: 0.9084
Epoch 10/10
1875/1875 [==============================] - 1s 674u

In [7]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.3443 - accuracy: 0.8773

Test accuracy: 0.8773000240325928


## Model #2 -  2. adam, dense10, dropout(.2), dense10

In [8]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(10, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [10]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 537us/step - loss: 0.9165 - accuracy: 0.6755
Epoch 2/10
1875/1875 [==============================] - 1s 528us/step - loss: 0.6876 - accuracy: 0.7508
Epoch 3/10
1875/1875 [==============================] - 1s 538us/step - loss: 0.6488 - accuracy: 0.7631
Epoch 4/10
1875/1875 [==============================] - 1s 551us/step - loss: 0.6205 - accuracy: 0.7727
Epoch 5/10
1875/1875 [==============================] - 1s 531us/step - loss: 0.6096 - accuracy: 0.7756
Epoch 6/10
1875/1875 [==============================] - 1s 538us/step - loss: 0.6018 - accuracy: 0.7778
Epoch 7/10
1875/1875 [==============================] - 1s 535us/step - loss: 0.5889 - accuracy: 0.7841
Epoch 8/10
1875/1875 [==============================] - 1s 537us/step - loss: 0.5852 - accuracy: 0.7850
Epoch 9/10
1875/1875 [==============================] - 1s 530us/step - loss: 0.5787 - accuracy: 0.7882
Epoch 10/10
1875/1875 [==============================] - 1s 534u

In [11]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.4604 - accuracy: 0.8357

Test accuracy: 0.8356999754905701


## Model #3: Model #1 performed better:  1. adagrad

In [12]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [13]:
model.compile(optimizer='adagrad',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [14]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 670us/step - loss: 1.0942 - accuracy: 0.6683
Epoch 2/10
1875/1875 [==============================] - 1s 634us/step - loss: 0.7261 - accuracy: 0.7679
Epoch 3/10
1875/1875 [==============================] - 1s 662us/step - loss: 0.6491 - accuracy: 0.7916
Epoch 4/10
1875/1875 [==============================] - 1s 631us/step - loss: 0.6084 - accuracy: 0.8032
Epoch 5/10
1875/1875 [==============================] - 1s 689us/step - loss: 0.5818 - accuracy: 0.8122
Epoch 6/10
1875/1875 [==============================] - 1s 637us/step - loss: 0.5626 - accuracy: 0.8166
Epoch 7/10
1875/1875 [==============================] - 1s 636us/step - loss: 0.5476 - accuracy: 0.8211
Epoch 8/10
1875/1875 [==============================] - 1s 642us/step - loss: 0.5357 - accuracy: 0.8246
Epoch 9/10
1875/1875 [==============================] - 1s 674us/step - loss: 0.5259 - accuracy: 0.8268
Epoch 10/10
1875/1875 [==============================] - 1s 637u

In [15]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.5382 - accuracy: 0.8208

Test accuracy: 0.8208000063896179


## Model #4: Model #1 performed better:  2. sgd

In [16]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10)
])

In [17]:
model.compile(optimizer='sgd',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

In [18]:
model.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 1s 616us/step - loss: 0.7348 - accuracy: 0.7614
Epoch 2/10
1875/1875 [==============================] - 1s 599us/step - loss: 0.5128 - accuracy: 0.8256
Epoch 3/10
1875/1875 [==============================] - 1s 601us/step - loss: 0.4679 - accuracy: 0.8398
Epoch 4/10
1875/1875 [==============================] - 1s 603us/step - loss: 0.4424 - accuracy: 0.8462
Epoch 5/10
1875/1875 [==============================] - 1s 603us/step - loss: 0.4249 - accuracy: 0.8519
Epoch 6/10
1875/1875 [==============================] - 1s 607us/step - loss: 0.4108 - accuracy: 0.8573
Epoch 7/10
1875/1875 [==============================] - 1s 598us/step - loss: 0.3988 - accuracy: 0.8621
Epoch 8/10
1875/1875 [==============================] - 1s 604us/step - loss: 0.3888 - accuracy: 0.8651
Epoch 9/10
1875/1875 [==============================] - 1s 602us/step - loss: 0.3795 - accuracy: 0.8670
Epoch 10/10
1875/1875 [==============================] - 1s 632u

In [19]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 0s - loss: 0.4076 - accuracy: 0.8551

Test accuracy: 0.8550999760627747


### Results

<p>Model #1<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.91151<br>
test acc 0.8773<br>
dense128, dense10 - same as original<br>

Model #2<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.7897<br>
test acc 0.8357<br>
dense10, dropout(.2), dense10<br>

Model #3<br>
options: relu, adagrad<br> 
train time: 1 second<br> 
train acc: 0.8291<br>
test acc 0.8208<br>


Model #4<br>
options: relu, sdg<br> 
train time: 1 second<br> 
train acc: 0.8705<br>
test acc 0.8551<br>